### DISCUSSION TITLE
Piecewise Linear Function for cost

### DISCUSSION THREAD

**Discussion by MatheusSuknaic** [google/or-tools#3386] [Routing (and legacy CP) questions]
I would like to know if in python is there a way to set a piecewise linear function to be the cost the model minimizes? 

For instance let's say we have one vehicle type, that if the route distance is one hundred meters it pays 0.80, if it's two hundred meters it pays 1.00, if it's three hundred meters it pays 1.30 dollars.

Basically, I want to change the evaluation function given that I have a route and not only change the arc costs of the matrix as `SetArcCostEvaluatorOfAllVehicles()` does.

**Comment by Mizux** [comment]
As you can see here, this code is SWIG guarded since we didn't have a working version of the PiecewiseLinearFunction so this method is
disabled in any wrapper...
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/routing.h#L2724-L2738

does this 0.80, 1.0 1.3 a fixed cost or the cost per km of the vehicle ?

**Comment by MatheusSuknaic** [comment]
This is a cost per km of the vehicle. Is there any other way around that I can model this?

**Reply by Mizux** [reply]
One workaround is to create three vehicles with a "gate" node.
- one vehicle V0 with 0.80 per km cost and a vehicle capacity of 100 with a fixed cost N
- one vehicle V1 with 1.00 per km cost and a vehicle capacity of 200 with a fixed cost N+
- one vehicle V2 with 1.30 per km cost and a vehicle capacity of 300 with a fixed cost N++

note1 : you must register three callback and use the AddDimension overload so you can pass an array of transit callback indices.

note2: Due to the fixed cost solver should have incentive to use the cheapest vehicle

- create a node at depot position all three vehicles's start node NextVar can only visit this one or go to end aka only one vehicle will have a route while two other will have no choice but to reach their end node.

```mermaid
graph LR;
v0S("V0 Start") ---> V0E("V0 End<br>(Max Cumul 100)") & G("Gate Node");
v1S("V1 Start") ---> V1E("V1 End<br>(Max Cumul 200)") & G;
v2S("V2 Start") ---> V2E("V2 End<br>(Max Cumul 300)") & G;
G ---> ...;
... ---> V0E & V1E & V2E
```
